app/controllers/pets_controller.rb

```diff
@@ -36,6 +40,28 @@ def destroy
     redirect_to '/pets'
   end
 
+  def favorite
+    pet = Pet.find(params[:id])
+    if session[:favorites] == nil
+      session[:favorites] = [params[:id]]
+      flash[:favorited] = " #{pet.name} has been added to your favorites!"
+    else
+      if session[:favorites].include?(params[:id])
+        session[:favorites].delete(params[:id])
+        flash[:favorited] = " #{pet.name} has been removed from your favorites!"
+      else
+        session[:favorites] << params[:id]
+        flash[:favorited] = "#{pet.name} has been added to your favorites!"
+      end
+    end
+    redirect_to "/pets/#{params[:id]}"
+  end
```

Comment:
Maybe look into a PORO for this?
The code and logic is great, but maybe it's too long of a method for the location